rust/src/librarydb.rs

```diff
@@ -0,0 +1,504 @@
+        }
+    }
+}
+
+/// Library only displays the path by default.
+impl fmt::Display for Library {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Library {{ {:?} }}", &self.library_path)
+    }
+}
+
+/// Library entries are ordered by file path.
+impl cmp::Ord for LibraryEntry {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.file_path.cmp(&other.file_path)
+    }
+}
+
+/// Library entries are ordered by file path.
+impl cmp::PartialOrd for LibraryEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Library entries are ordered by file path.
+impl cmp::PartialEq for LibraryEntry {
+    fn eq(&self, other: &Self) -> bool {
+        self.file_path.eq(&other.file_path)
+    }
+}
+
+impl cmp::Eq for LibraryEntry {}
+
+/// LibraryFile seeks the data of a library entry.
+impl io::Seek for LibraryFile {
+    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+        let library = self.arc_library.write().unwrap();
+        let entry = &library.entries[self.index];
+        let checked_position = match pos {
+            SeekFrom::Start(n) => entry.data_start.checked_add(n),
+            SeekFrom::Current(n) => checked_add_u64_i64(self.position, n),
+            SeekFrom::End(n) => checked_add_u64_i64(entry.data_end, n),
+        };
+        if let Some(position) = checked_position {
+            if position >= entry.data_start {
+                self.position = position;
+                return Ok(position - entry.data_start);
+            }
+        }
+        // must never become negative or overflow
+        Err(io::ErrorKind::InvalidInput.into())
+    }
+}
+
+/// LibraryFile reads the data of a library entry.
+impl io::Read for LibraryFile {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let mut library = self.arc_library.write().unwrap();
+        let end = library.entries[self.index].data_end;
+        if self.position >= end {
+            return Ok(0);
+        }
+        let available = end - self.position;
+        library.library_file.seek(SeekFrom::Start(self.position))?;
+        let read_result = if available < buf.len() as u64 {
+            library
+                .library_file
+                .read(buf.split_at_mut(available as usize).0)
+        } else {
+            library.library_file.read(buf)
+        };
+        if let Ok(bytes) = read_result {
+            self.position += bytes as u64;
+        }
+        read_result
+    }
+}
+
+/// Performs a checked unsigned signed addition.
+fn checked_add_u64_i64(a: u64, b: i64) -> Option<u64> {
+    if b > 0 {
+        a.checked_add(b as u64)
+    } else if b < 0 {
+        a.checked_sub(-b as u64)
+    } else {
+        Some(a)
+    }
+}
+
+/// Gets the case insensitive representation of a path in the library database.
+fn case_insensitive_path(path: &str) -> String {
+    path.to_ascii_uppercase().replace('\\', "/")
```

Comment:
Because unicode is [complicated](https://www.w3.org/International/wiki/Case_folding)... a true case insensitive comparison is not just converting to uppercase/lowercase and comparing.

I'm not confident that my implementation would be correct in all cases, so it's better to stick to ascii until rust can do case insensitive comparisons.